### DISCUSSION TITLE
Registration on batches of images?

### DISCUSSION THREAD

**Discussion by carbocation** [kornia/kornia#2541] [General]
The registrator.register() method is pretty clear that it is designed to work on an image (signature and docstring below). However, it would be optimal for my use case to be able to apply this to a batch of images. I thought this might be possible, since we have to manipulate our input so that the batch size is 1 for both the source and the destination. But when I expand that (e.g., batch size of 50), I get an error, also pasted below. This isn't totally surprising, but I guess since we had to provide a batch size I thought perhaps we could make this work on batches greater than size 1. Has anyone tried or succeeded in doing registration with kornia on multiple images at once?

```py
def register(
        self, src_img: Tensor, dst_img: Tensor, verbose: bool = False, output_intermediate_models: bool = False
    ) -> Union[Tensor, Tuple[Tensor, List[Tensor]]]:
        r"""Estimate the tranformation' which warps src_img into dst_img by gradient descent. The shape of the
        tensors is not checked, because it may depend on the model, e.g. volume registration.

        Args:
            src_img: Input image tensor.
            dst_img: Input image tensor.
            verbose: if True, outputs loss every 10 iterations.
            output_intermediate_models: if True with intermediate models

        Returns:
            the transformation between two images, shape depends on the model,
            typically [1x3x3] tensor for string model_types.
        """
```

Input shapes:
```py
src:  torch.Size([50, 3, 204, 208])
dest: torch.Size([50, 3, 204, 208])
```

Error from batch size 50:

```py
Traceback (most recent call last):
  File "registration.py", line 148, in <module>
    transformations = registrator.register(src_img = rgb_image_tensor[0], dst_img = dest_image, verbose=True)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/kornia/geometry/transform/image_registrator.py", line 240, in register
    loss = self.get_single_level_loss(img_src_level, img_dst_level, self.model())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/kornia/geometry/transform/image_registrator.py", line 189, in get_single_level_loss
    raise ValueError(
ValueError: Cannot register images of different shapes                              torch.Size([50, 3, 12, 13]) torch.Size([1, 3, 12, 13])
```

**Comment by edgarriba** [comment]
@carbocation are you batch elements map 1->1 ? in any case, the algorithm is kinda greedy that every iteration has a stop condition. To solve by batch you should have an optimizer or parallelize somehow. @ducha-aiki

**Reply by ducha-aiki** [reply]
Yes, the stopping condition is a problem here. However, if you modify the transformation class to accept batches, it probably should work with small modifications.
